homeassistant/components/lamarzocco/calendar.py

```diff
@@ -0,0 +1,99 @@
+"""Calendar platform for La Marzocco espresso machines."""
+
+from collections.abc import Iterator
+from datetime import datetime, timedelta
+
+from homeassistant.components.calendar import CalendarEntity, CalendarEvent
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util import dt as dt_util
+
+from .const import DOMAIN
+from .entity import LaMarzoccoBaseEntity
+
+CALENDAR_KEY = "auto_on_off_schedule"
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up switch entities and services."""
+
+    coordinator = hass.data[DOMAIN][config_entry.entry_id]
+    async_add_entities([LaMarzoccoCalendarEntity(coordinator, CALENDAR_KEY)])
+
+
+class LaMarzoccoCalendarEntity(LaMarzoccoBaseEntity, CalendarEntity):
+    """Class representing a La Marzocco calendar."""
+
+    _attr_translation_key = CALENDAR_KEY
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the next upcoming event."""
+        # only need to check the next 6 days, because if we don't find anything there
+        # then there is no event scheduled
+        now = dt_util.now()
+        for date in self._get_date_range(now, now + timedelta(days=6)):
+            if scheduled := self._async_get_calendar_event(date):
+                if scheduled.end < now:
+                    continue
+                return scheduled
+        return None
+
+    async def async_get_events(
```

Comment:
I think it may be possible for this to skip events in the range.

Say you have a schedule of: "Every Sunday between 07:00 - 07:30 (only)" then this function is called with:
- start: 2024-02-11 09:00
- end: 2024-02-18 08:00

Then I *think* it may not return the event on 2024-02-18 07:00 but it might return the event on 2024-02-11 07:00 ? I think this happens because it gets the days involved before checking the schedule.

I imagine this doesn't have a big impact in practice for the calendar UI based on how this is called, but i could see this being a problem for triggers where this gets called with a small window and if its not correct then nothing will get triggered and its hard to debug. 

(For what it's worth, I have a very over designed solution to this problem for [rainbird sprinkler](https://github.com/allenporter/pyrainbird/blob/440684f238f6f82d47de9e3dd2642f81adf782f8/pyrainbird/timeline.py#L79) using the same underlying libraries used by local calendar to generate a timeline based on a schedule and recurrence rules. You don't need to use this or anything, but i was just sharing since i ran into a similar problem there )

Reply:
The only thing that could've happened, is that we missed any events on the same weekday in the following week that were already ended at the current time. E.g. it is Sunday 08:01 and the next event is next Sunday 07:30 - 8:00 we would've missed that. I added an extra day to the date range, that should fix that problem. If the event is only at 08:30 we would get it twice, but since we only take the next occurence here, that's no problem.

Reply:
The other thing to be careful of is returning an event before the range though too and omit those.

This may not cause a major problem for this integration since it's once per day, but in general it will also screw up triggers since it assumes it can correctly iterate  through events by scanning the time range and see them only when they are supposed to be seen -- and I don't want to have that pattern copied into other calendars.

Reply:
Such an example is part of the tests (Friday iirc)

Reply:
I am referring to the example in my comment. The start datetime is on a day present in the schedule and an event is returned outside of the range. My impression is the steps are this:
- Call `async_get_events(2024-02-11 09:00, 2024-02-18 08:00)`
- Call `_get_date_range(2024-02-11 09:00, 2024-02-18 08:00)`
- _get_date_range yields `2024-02-11 09:00`
- Call `_async_get_calendar_event(2024-02-11 09:00`
- date.weekday is sunday
- returns `CalendarEvent(start=2024-02-11 07:00, end=2024-02-11 07:30`
Maybe this is prevented and i'm missing it.

Reply:
ah, I thought you were talking about the `property` and not `async_get_events`. Indeed that was possible (and actually I had exactly such a case in my snapshot test). Moved the logic to a separate function, that should be caught now.

Reply:
Yes i'm talking about `async_get_events` which has a start and end range and is what is called by the trigger.

I still think there are some cases missed by the current logic.  Given a schedule of "Every Sunday between 07:00 - 07:30", I don't think all of these cases are handled correctly:
- start: 2024-02-11 06:00, end: 2024-02-18 06:00 => Includes 2024-02-12 07:00 only
- start: 2024-02-11 07:15, end: 2024-02-18 06:00 => Includes 2024-02-12 07:00 only
- start: 2024-02-11 09:00, end: 2024-02-18 07:15 => Includes 2024-02-18 07:00 only
- start: 2024-02-11 09:00, end: 2024-02-18 08:00 => Includes 2024-02-18 07:00 only
- start: 2024-02-11 09:00, end: 2024-02-18 06:00 => Includes none
- start: 2024-02-11 06:00, end: 2024-02-18 08:00 => Includes both

I think the current logic will still include events outside the range when the end date is a day that has an event scheduled, but the calendar event is outside the range. I think its because `_get_events` needs another check like the start date check but inverted for the end date check.

A parameterized test case can be a straight forward way to exercise all of these cases with a single test

Reply:
I have to say, I'm impressed that you caught those without running the code! Indeed, there were a few of those which were wrong. I fixed them and added a test case as suggested.

Reply:
And thanks for the detailed explanation!